### PR TITLE
ASoC: SOF: Intel: hda: refine resume with init chip

### PR DESCRIPTION
**Pull request by zhuyingjiang**
Unify resume code by using SOF common function hda_dsp_ctrl_init_chip() which can handle both HDA and
non-HDA cases.

### CODE REVIEW

**Comment by plbossart** [comment]
I don't have time to look into this so will have to rely on others to approve.

**Comment by ranj063** [comment]
@zhuyingjiang while the overall flow looks OK, can you please check why you need to clear WAKESTS twice in hda_dsp_ctrl_init_chip() for the HDA case? 

Also, clearing WAKESTS should use snd_sof_dsp_write, update_bits() will not really clear it.

**Comment by plbossart** [comment]
@zhuyingjiang also please resolve conflicts

**Comment by zhuyingjiang** [comment]
> Also, clearing WAKESTS should use snd_sof_dsp_write, update_bits() will not really clear it.

@ranj063 As I checked, here should use snd_sof_dsp_write, I will update it.

**Comment by zhuyingjiang** [comment]
> Code looks good, but again a few notes to the commit message. See also Ranjani's note on WAKESTS clearing.
> 
> Commit message for first -- minor change to make it easier to read.
> 
> ```
> Set the HDA stream position buffer during init chip. The position buffer
> needs to be set in both HDA codec and nocodec cases. Use the SOF
> defined function and move it to common code.
> ```
> 
> Commit message for the second -- here the explanation goes in a weird
> order. I think it's better to start with the "unify resume code" to explain why
> you are doing this and then add detail about the implementation. Otherwise
> it's a bit hard to follow if you start with details and then reveal at the end
> why you did it.
> 
> ```
> Unify resume code by using SOF common function hda_dsp_ctrl_init_chip()
> which can handle both HDA and non-HDA cases. Move code to reset
> stream-to-link mapping into hda_dsp_ctrl_init_chip.
> ```
> 
> Also in both, you can word-wrap to 74 columns.

Thanks @kv2019i Will update it after some other check.

**Comment by zhuyingjiang** [comment]
@plbossart @kv2019i @ranj063 Updated.

**Comment by ranj063** [comment]
> Thanks, commits look good!
> 
> @RanderWang , @lyakh and @ranj063 : just to double check, is is safe to not do WAKESTS clearing before controller reset? The old code was wrong (it should not use update bits), but given we had many issues with the headset/wake detection, I want to be extra careful here.

@kv2019i I have a feeling this should be fine. With WAKEEN , we could be resuming because of a codec status change which is recorded before the controller is reset. But let me test this PR today just to make sure

**Comment by zhuyingjiang** [comment]
> Thanks, commits look good!
> 
> @RanderWang , @lyakh and @ranj063 : just to double check, is is safe to not do WAKESTS clearing before controller reset? The old code was wrong (it should not use update bits), but given we had many issues with the headset/wake detection, I want to be extra careful here.

@kv2019i  I have already test this on WHL RVP. It is OK.
The register write steps are kept the same as other drivers, there clear is called multi-times in different functions.
Here we combine them in one function. and As WAKESTS resides in the Primary well, only one clear is OK for now.

**Comment by kv2019i** [comment]
@zhuyingjiang 
> @kv2019i I have already test this on WHL RVP. It is OK.

Ok great, thanks for confirming. Then this looks good to go.

**Comment by ranj063** [comment]
@zhuyingjiang can you please update your branch ?

**Comment by zhuyingjiang** [comment]
> @zhuyingjiang can you please update your branch ?

@ranj063  Sorry I don't get what you mean by "update branch", the target branch is not sof-dev? or there still need some more modification?

**Comment by ranj063** [comment]
> > @zhuyingjiang can you please update your branch ?
> 
> @ranj063 Sorry I don't get what you mean by "update branch", the target branch is not sof-dev? or there still need some more modification?

@zhuyingjiang the status shows that the branch is out-of-date with the base branch. So I cant merge it as is

**Comment by zhuyingjiang** [comment]
@ranj063 @keyonjie  Updated.

**Comment by zhuyingjiang** [comment]
@kv2019i Updated the base branch, no code and commit message changes, could you review and approve again?

File: sound/soc/sof/intel/hda-ctrl.c [hunk 1]
```diff
@@ -176,11 +176,6 @@ int hda_dsp_ctrl_init_chip(struct snd_sof_dev *sdev, bool full_reset)
 	hda_dsp_ctrl_misc_clock_gating(sdev, false);
 
 	if (full_reset) {
-		/* clear WAKESTS */
-		snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR, SOF_HDA_WAKESTS,
-					SOF_HDA_WAKESTS_INT_MASK,
-					SOF_HDA_WAKESTS_INT_MASK);
-
```
Comment (keyonjie) [inline sound/soc/sof/intel/hda-ctrl.c:180]: should remove the later one, and keep this and move it out of if (full_reset), above.

File: sound/soc/sof/intel/hda-ctrl.c [hunk 2]
```diff
@@ -164,6 +164,9 @@ int hda_dsp_ctrl_clock_power_gating(struct snd_sof_dev *sdev, bool enable)
 int hda_dsp_ctrl_init_chip(struct snd_sof_dev *sdev, bool full_reset)
 {
 	struct hdac_bus *bus = sof_to_bus(sdev);
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+	struct hdac_ext_link *hlink = NULL;
```
Comment (lyakh) [inline sound/soc/sof/intel/hda-ctrl.c]: Looks like a superfluous initialisation, please, check and - if confirmed - remove

Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-ctrl.c]: @lyakh the code is needed for reset the stream-to-link mapping, both for chip power up and the DSP resume.

File: sound/soc/sof/intel/hda-ctrl.c (hunk 1 above)
Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-ctrl.c:180]: @keyonjie You are partly right, keep this one and remove the later one does work. 

but move it out will leads to problems:
if full_reset is false, that means the wake status cleared but no actual reset.
because after reset, the wake status register will be set by the hardware and the codec_mask is from there.

So I will change the PR to: 
clear the WAKESTS and reset , also set codec_mask to 0; then after reset, the codec_mask will read  again from the register. this will handle the case when suspend, and then the codec has been changed .

**Comment by zhuyingjiang** [comment]
> Hi @zhuyingjiang , can you try it more aggressive, to remove using many "#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)"? Something like this:

@keyonjie  we must follow the steps, that is first reset and clear then enable. if move code like that, it will leads to reset, enable and then clear. this will leads to problems.

File: sound/soc/sof/intel/hda-ctrl.c (hunk 1 above)
Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-ctrl.c:180]: @keyonjie A update: the later one can't be removed, or the device can't suspend after the first suspend.

File: sound/soc/sof/intel/hda-ctrl.c [hunk 3]
```diff
@@ -177,9 +177,14 @@ int hda_dsp_ctrl_init_chip(struct snd_sof_dev *sdev, bool full_reset)
...
-					SOF_HDA_WAKESTS_INT_MASK,
-					SOF_HDA_WAKESTS_INT_MASK);
+		snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR, SOF_HDA_WAKESTS,
+				  SOF_HDA_WAKESTS_INT_MASK);
+
+		/*
+		 * set codec_mask to 0, and get it again after reset. This
+		 * will handle the case that codecs changed during suspend.
+		 */
+		bus->codec_mask = 0;
```
Comment (RanderWang) [inline sound/soc/sof/intel/hda-ctrl.c]: This codec_mask is reset by codec driver in hdac_hda.c and hdmi driver in hdac_hdmi.c after it is initialized by SOF.
According to legacy hda driver, these operations are not reasonable in ASoC.

This patch is a workaround for above case. I recommend doesn't change it now. 
codec_mask is only for skl_hda_generic driver to initialize dai link when PCI probed, and it is not used later.
So driver works without this patch.

Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-ctrl.c]: @RanderWang @keyonjie @ranj063 So we are back to the beginning.
Just remove the first clear WAKESTS, and do nothing else in the last commit.

Comment (ranj063) [reply sound/soc/sof/intel/hda-ctrl.c]: @zhuyingjiang I'm confused about  your comment and the commit. Do you have to set codec_mask to 0 here?

Comment (kv2019i) [reply sound/soc/sof/intel/hda-ctrl.c]: @zhuyingjiang Maybe better to leave this out. I agree with @ranj063 that the comment+commit are not clear. Unless there's a good reason, I'd keep the current logic where we read codec_mask once after probe with:
```
»       if (!bus->codec_mask) {
»       »       bus->codec_mask = snd_hdac_chip_readw(bus, STATESTS);
```

... and not reset unless probe is done again (by unloading and reloading the modules).

Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-ctrl.c]: @kv2019i That is what I say we are back to the beginning. So I will push the old version of my patch.

Why set codec_mask to 0 is because I see an issue that with a printk, I see codec_mask is 0x5 at cold start, and after suspend/resume, it change to 0x1. I set codec_mask to 0 to make the codec_mask to be read again from the STATESTS, (will got 0x5).

But as @RanderWang suggested 
> This patch is a workaround

we need to fix the above issue elsewhere, maybe i915 issue or somewhere else.

And what @keyonjie suggested:
> should remove the later one, and keep this and move it out of if (full_reset), above.

As I tested on WHL RVP,  the later one can't be removed, or the device can't suspend after the first suspend.

In conclude: now we are aligned that we only need remove the first 'remove STATESTS' , that was what I commit in the previous version.

I will push the previous version again.

**Comment by zhuyingjiang** [comment]
@kv2019i @lyakh @ranj063 @keyonjie @RanderWang Now we are aligned, pushed the old version again.
And the Jack detection is also tested, can detect jack plugging in.

**Comment by ranj063** [comment]
@zhuyingjiang can you please rebase and resubmit. It still shows that the branch is out of date

**Comment by zhuyingjiang** [comment]
> @zhuyingjiang can you please rebase and resubmit. It still shows that the branch is out of date

@ranj063 @RanderWang @keyonjie Rebased.

**Comment by zhuyingjiang** [comment]
@lyakh

**Comment by zhuyingjiang** [comment]
@lyakh Your requested changes need to be addressed to merge the PR. Please review for me, thanks!

File: sound/soc/sof/intel/hda-ctrl.c (hunk 2 above)
Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-ctrl.c]: @lyakh removed that initialisation.

**Comment by zhuyingjiang** [comment]
@lyakh removed that initialization.

**Comment by ranj063** [comment]
@zhuyingjiang sorry I've just merged the upstream changes. So could you please rebase and resubmit?

**Comment by zhuyingjiang** [comment]
> @zhuyingjiang sorry I've just merged the upstream changes. So could you please rebase and resubmit?

OK.

**Comment by zhuyingjiang** [comment]
@ranj063 @lyakh Rebased.
